### PR TITLE
KAFKA-14616: Fix stray replica of recreated topics in KRaft mode

### PR DESCRIPTION
**Pull request by cmccabe**
When a broker is down, and a topic is deleted, this will result in that broker seeing "stray replicas" the next time it starts up. These replicas contain data that used to be important, but which now needs to be deleted. Stray replica deletion is handled during the initial metadata publishing step on the broker.

Previously, we deleted these stray replicas after starting up BOTH LogManager and ReplicaManager. However, this wasn't quite correct. The presence of the stray replicas confused ReplicaManager. Instead, we should delete the stray replicas BEFORE starting ReplicaManager.

This bug triggered when a topic was deleted and re-created while a broker was down, and some of the replicas of the re-created topic landed on that broker. The impact was that the stray replicas were deleted, but the new replicas for the next iteration of the topic never got created. This, in turn, led to persistent under-replication until the next time the broker was restarted.

### CODE REVIEW

**Review by gnarula — APPROVED** [review]
Thanks for the fix! LGTM!

Tested the fix on 3.7 branch (at `f192435`) with the following steps:

* Created a cluster with 3 controllers, 4 brokers, 1 disk in each
* Created a topic, `test-topic` with 10 partitions and 4 replicas each. The topic description was as follows:

```
Topic: test-topic       TopicId: e4rNbTcjT_elpYRxU0Jgtw PartitionCount: 10      ReplicationFactor: 4    Configs: min.insync.replicas=1,unclean.leader.election.enable=false
        Topic: test-topic       Partition: 0    Leader: 103     Replicas: 103,104,101,102       Isr: 103,104,101,102
        Topic: test-topic       Partition: 1    Leader: 104     Replicas: 104,101,102,103       Isr: 104,101,102,103
        Topic: test-topic       Partition: 2    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103,104
        Topic: test-topic       Partition: 3    Leader: 102     Replicas: 102,103,104,101       Isr: 102,103,104,101
        Topic: test-topic       Partition: 4    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103,104
        Topic: test-topic       Partition: 5    Leader: 102     Replicas: 102,103,104,101       Isr: 102,103,104,101
        Topic: test-topic       Partition: 6    Leader: 103     Replicas: 103,104,101,102       Isr: 103,104,101,102
        Topic: test-topic       Partition: 7    Leader: 104     Replicas: 104,101,102,103       Isr: 104,101,102,103
        Topic: test-topic       Partition: 8    Leader: 102     Replicas: 102,104,103,101       Isr: 102,104,103,101
        Topic: test-topic       Partition: 9    Leader: 104     Replicas: 104,103,101,102       Isr: 104,103,101,102
```

* Brought `broker-4` down.
* Deleted `test-topic` and recreated it. The topic description was:
```
Topic: test-topic       TopicId: w8vJ34nSRxqf_5zZCr8Pbg PartitionCount: 10      ReplicationFactor: 4    Configs: min.insync.replicas=1,unclean.leader.election.enable=false
        Topic: test-topic       Partition: 0    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103
        Topic: test-topic       Partition: 1    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101
        Topic: test-topic       Partition: 2    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102
        Topic: test-topic       Partition: 3    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102
        Topic: test-topic       Partition: 4    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103
        Topic: test-topic       Partition: 5    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101
        Topic: test-topic       Partition: 6    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103
        Topic: test-topic       Partition: 7    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101
        Topic: test-topic       Partition: 8    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102
        Topic: test-topic       Partition: 9    Leader: 103     Replicas: 103,102,101,104       Isr: 103,102,101
```
Notice that the topic id is new and ISR doesn't include broker-4 because it is down

* Brought broker-4 online. Observed the broker marked previous replicas as stray and created replicas for new topic id

```
Topic: test-topic       TopicId: w8vJ34nSRxqf_5zZCr8Pbg PartitionCount: 10      ReplicationFactor: 4    Configs: min.insync.replicas=1,unclean.leader.election.enable=false
        Topic: test-topic       Partition: 0    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103,104
        Topic: test-topic       Partition: 1    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101,104
        Topic: test-topic       Partition: 2    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102,104
        Topic: test-topic       Partition: 3    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102,104
        Topic: test-topic       Partition: 4    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103,104
        Topic: test-topic       Partition: 5    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101,104
        Topic: test-topic       Partition: 6    Leader: 101     Replicas: 101,102,103,104       Isr: 101,102,103,104
        Topic: test-topic       Partition: 7    Leader: 102     Replicas: 102,103,101,104       Isr: 102,103,101,104
        Topic: test-topic       Partition: 8    Leader: 103     Replicas: 103,101,102,104       Isr: 103,101,102,104
        Topic: test-topic       Partition: 9    Leader: 103     Replicas: 103,102,101,104       Isr: 103,102,101,104
```

```
ls -alh .cache/data/broker-4/d1/
Permissions Size User   Date Modified Name
.rw-r--r--     0 gaurav 24 Jan 14:51  .lock
.rw-r--r--   249 gaurav 24 Jan 14:51  bootstrap.checkpoint
.rw-r--r--     0 gaurav 24 Jan 14:51  cleaner-offset-checkpoint
.rw-r--r--     4 gaurav 24 Jan 14:54  log-start-offset-checkpoint
.rw-r--r--   124 gaurav 24 Jan 14:51  meta.properties
.rw-r--r--   155 gaurav 24 Jan 14:54  recovery-point-offset-checkpoint
.rw-r--r--   160 gaurav 24 Jan 14:54  replication-offset-checkpoint
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-0
drwxr-xr-x     - gaurav 24 Jan 14:52  test-topic-0.c5bc6e18cae140aa88f698ea2ba119f7-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-1
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-1.e48f37ad0d7c403c82b3e74b4c2811dc-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-2
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-2.c07070fd3d9d4b6282ee72ac6b3113f0-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-3
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-3.9b5b16ac05ac49fab26e0762d60dcad1-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-4
drwxr-xr-x     - gaurav 24 Jan 14:52  test-topic-4.eea62fcfe5b547c9bf4ebdfd9d5e2299-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-5
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-5.d1e59fc3d1de40e8b02d73a8035a67ce-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-6
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-6.d940acfadf6442a2a81f654c34afa798-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-7
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-7.90c7161f8d2a40ceb2a2accb646c2b46-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-8
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-8.544860cc3b7c4d70abd8c5f74f90f717-stray
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-9
drwxr-xr-x     - gaurav 24 Jan 14:53  test-topic-9.1761f1225e4641a8b6471e1ab3fcf787-stray
```

```
cat .cache/data/broker-4/d1/test-topic-0/partition.metadata
version: 0
topic_id: w8vJ34nSRxqf_5zZCr8Pbg
```

```
cat .cache/data/broker-4/d1/test-topic-0.c5bc6e18cae140aa88f698ea2ba119f7-stray/partition.metadata
version: 0
topic_id: e4rNbTcjT_elpYRxU0Jgtw
```

File: core/src/main/scala/kafka/log/LogManager.scala
Comment (showuon) [inline core/src/main/scala/kafka/log/LogManager.scala:576]: 1. We should not `Reinitialize` the metadata since the stray partitions are unused anymore.
2. The renameDir is already done in `asyncDelete` if `isStray` is true, which is what we have now. So we don't need to renameDir here.

**Review by showuon — COMMENTED** [review]
Overall LGTM! Just one comment. Thanks.

**Comment by showuon** [comment]
Question: why can't we delete these disappeared(deleted) logs directly? I think this can only happen when topic deleting while the node is offline. If so, then deleting them should be fine?

**Comment by OmniaGM** [comment]
> Question: why can't we delete these disappeared(deleted) logs directly? I think this can only happen when topic deleting while the node is offline. If so, then deleting them should be fine?

The bug only happened when we delete and recreate the topic while the broker is offline and there is overlap on partition assignment between the new and old topic on the offline broker. 

A better way to explain in it is by the example I mentioned in the Jira where 
1. we had offline broker that held replicas for partitions  0, 3, 4, 5, 7, 8, 9 of topic `foo.test` with id `MfuZbwdmSMaiSa0g6__TPg` which we deleted while the broker is offline so they aren't physically deleted. 
2. Then before brining the offline broker we recreated `foo.test` again which now have partitions 0, 1, 2, 7, 8, 9 assigned to same offline broker but now the topic id is `RzalpqQ9Q7ub2M2afHxY4Q`. **Notice here that partitions 0, 7, 8, 9 are common between the assignment of the deleted topic (topic_id: MfuZbwdmSMaiSa0g6__TPg) and the recreated topic (topic_id: RzalpqQ9Q7ub2M2afHxY4Q).**
3. When the broker come-back the `ReplicaManager` and `LogManager` will not complain and move one to create the missing partitions' log dir for partitions 1, 2, 6 as there are already dirs for 0, 7, 8, 9 (despite the fact that the topic id is wrong here). 
4. Then later `BrokerMetadataPublisher` will detect the strays (which are 0, 7, 8, 9) and deleted them. 
5. Now the broker is online but has only partition 1, 2, 6 and we have no other way to bring the broker to create partition 0, 7, 8, 9 without restarting the broker. 

which lead to permanent under-replication until someone restart the broker. It could get worse on larger clusters if we have more than 1 offline broker and the same issue happened to all of them. We might end up with partition in permeant at-min-isr or permeant under-min-isr.

**Review by OmniaGM — APPROVED** [review]
Thanks for the pr! The fix LGTM!

File: core/src/main/scala/kafka/log/LogManager.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/log/LogManager.scala:576]: OK. I will delete the call to `log.renameDir`.

**Comment by cmccabe** [comment]
> Question: why can't we delete these disappeared(deleted) logs directly? I think this can only happen when topic deleting while the node is offline. If so, then deleting them should be fine?

It probably would be fine, but I'm a bit nervous about it. What I'd really like is to delete them after a time delay of a few hours, but the code doesn't support that currently. We should add this functionality.

**Comment by cmccabe** [comment]
committed

**Comment by mimaison** [comment]
When not merging via github we should include the PR number in the commit message otherwise the PR and commit are not linked. When digging through history it's often important to find the PR and discussion to understand some changes.

Commit: https://github.com/apache/kafka/commit/f7feb43af351efa25be37cb78814c19bd28a22a2

File: core/src/main/scala/kafka/log/LogManager.scala
Comment (showuon) [reply core/src/main/scala/kafka/log/LogManager.scala:576]: I didn't see this change. Did you miss it?

**Review by showuon — COMMENTED** [review]
@cmccabe , since there's still outstanding comment you haven't addressed, it would be better if you could wait until reviewer's approval before merge it. Thank you.

**Pull request closed without merging**
